### PR TITLE
perf(lua): optimize vim.deep_equal

### PR DESCRIPTION
**Pull request by muniter**
By remembering the keys already compared in a, repeating comparisons is avoided. Thanks: https://stackoverflow.com/a/32660766

I've also added documentation.

### CODE REVIEW

File: runtime/lua/vim/shared.lua [hunk 1]
```diff
@@ -267,18 +267,25 @@ function vim.tbl_deep_extend(behavior, ...)
...
     for k, v in pairs(a) do
       if not vim.deep_equal(v, b[k]) then
         return false
       end
+      keyset[k] = true
     end
-    for k, v in pairs(b) do
-      if not vim.deep_equal(v, a[k]) then
+    for k, _ in pairs(b) do
+      if not keyset[k] then
```
Comment (bfredl) [inline runtime/lua/vim/shared.lua:288]: Can't this just be `a[k] ~= nil` and eliminate `keyset` ?

Comment (muniter) [reply runtime/lua/vim/shared.lua:288]: Yes that would be the same! Will change it.

File: runtime/lua/vim/shared.lua [hunk 2]
```diff
@@ -267,18 +267,23 @@ function vim.tbl_deep_extend(behavior, ...)
 end
 
 --- Deep compare values for equality
+---
+--- Tables are compared recursively unless they both provide the `eq` methamethod.
+--- All other types are compared using the equality `==` operator.
+--@param a first value
+--@param b second value
+--@returns `true` if values are equals, else `false`.
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua:275]: format was updated, see `:help dev-lua-doc`

```suggestion
---@param a first value
---@param b second value
---@returns `true` if values are equals, else `false`.
```

File: runtime/doc/lua.txt [hunk 3]
```diff
@@ -1210,7 +1210,17 @@ schedule_wrap({cb})                                      *vim.schedule_wrap()*
...
+
+                Tables are compared recursively unless they both provide the `eq` methamethod.
+                All other types are compared using the equality `==` operator.
+
+                Parameters: ~
+                    {a}  first value
+                    {b}  second value
+
+                Return: ~
+                    `true` if values are equals, else `false` .
```
Comment (justinmk) [inline runtime/doc/lua.txt:1223]: future note: this is auto-generated

File: runtime/lua/vim/shared.lua [hunk 4]
```diff
@@ -267,18 +267,23 @@ function vim.tbl_deep_extend(behavior, ...)
 end
 
 --- Deep compare values for equality
+---
+--- Tables are compared recursively unless they both provide the `eq` methamethod.
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua:271]: weird that we named this `deep_equal` instead of `eq_something` for parallel with Lua's convention

**Pull request merged**